features/lorawan/lorawan_types.h

```diff
@@ -366,4 +367,60 @@ typedef struct lora_channelplan {
     loramac_channel_t *channels;
 } lorawan_channelplan_t;
 
+/**
+ * Meta-data collection for a transmission
+ */
+typedef struct {
+    /**
+     * A boolean to mark if the meta data is stale
+     */
+    bool stale;
+    /**
+     * The uplink channel used for transmission.
+     */
+    uint32_t channel;
+    /**
+     * The uplink datarate.
+     */
+    uint8_t data_rate;
+    /**
+     * The transmission power.
+     */
+    int8_t tx_power;
+    /**
+     * Provides the number of retransmissions.
+     */
+    uint8_t nb_retries;
+    /**
+     * The transmission time on air of the frame.
+     */
+    uint32_t tx_toa;
+} lorawan_tx_metadata;
+
+/**
+ * Meta-data collection for the received packet
+ */
+typedef struct {
+    /**
+     * A boolean to mark if the meta data is stale
+     */
+    bool stale;
+    /**
+     * Data rate of reception
+     */
+    uint8_t rx_datarate;
+    /**
+     * Frame pending status.
+     */
+    uint8_t fpending_status;
```

Comment:
Probably fpending bit is not needed. If auto uplink is disabled, stack sends UPLINK_REQUIRED event.